osquery/core/hashing.cpp

```diff
@@ -0,0 +1,142 @@
+
+namespace osquery {
+
+/// The buffer read size from file IO to hashing structures.
+const size_t kHashChunkSize{4096};
+
+Hash::~Hash() {
+  if (ctx_ != nullptr) {
+    free(ctx_);
+  }
+}
+
+Hash::Hash(HashType algorithm) : algorithm_(algorithm) {
+  if (algorithm_ == HASH_TYPE_MD5) {
+    length_ = MD5_DIGEST_LENGTH;
+    ctx_ = static_cast<MD5_CTX*>(malloc(sizeof(MD5_CTX)));
+    MD5_Init(static_cast<MD5_CTX*>(ctx_));
+  } else if (algorithm_ == HASH_TYPE_SHA1) {
+    length_ = SHA_DIGEST_LENGTH;
+    ctx_ = static_cast<SHA_CTX*>(malloc(sizeof(SHA_CTX)));
+    SHA1_Init(static_cast<SHA_CTX*>(ctx_));
+  } else if (algorithm_ == HASH_TYPE_SHA256) {
+    length_ = SHA256_DIGEST_LENGTH;
+    ctx_ = static_cast<SHA256_CTX*>(malloc(sizeof(SHA256_CTX)));
+    SHA256_Init(static_cast<SHA256_CTX*>(ctx_));
+  } else {
+    throw std::domain_error("Unknown hash function");
+  }
+}
+
+void Hash::update(const void* buffer, size_t size) {
+  if (algorithm_ == HASH_TYPE_MD5) {
+    MD5_Update(static_cast<MD5_CTX*>(ctx_), buffer, size);
+  } else if (algorithm_ == HASH_TYPE_SHA1) {
+    SHA1_Update(static_cast<SHA_CTX*>(ctx_), buffer, size);
+  } else if (algorithm_ == HASH_TYPE_SHA256) {
+    SHA256_Update(static_cast<SHA256_CTX*>(ctx_), buffer, size);
+  }
+}
+
+std::string Hash::digest() {
+  std::vector<unsigned char> hash;
+  hash.assign(length_, '\0');
+
+  if (algorithm_ == HASH_TYPE_MD5) {
+    MD5_Final(hash.data(), static_cast<MD5_CTX*>(ctx_));
+  } else if (algorithm_ == HASH_TYPE_SHA1) {
+    SHA1_Final(hash.data(), static_cast<SHA_CTX*>(ctx_));
+  } else if (algorithm_ == HASH_TYPE_SHA256) {
+    SHA256_Final(hash.data(), static_cast<SHA256_CTX*>(ctx_));
+  }
+
+  // The hash value is only relevant as a hex digest.
+  std::stringstream digest;
+  for (size_t i = 0; i < length_; i++) {
+    digest << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
+  }
+
+  return digest.str();
+}
+
+std::string hashFromBuffer(HashType hash_type,
+                           const void* buffer,
+                           size_t size) {
+  Hash hash(hash_type);
+  hash.update(buffer, size);
+  return hash.digest();
+}
+
+MultiHashes hashMultiFromFile(int mask, const std::string& path) {
+  std::map<HashType, std::shared_ptr<Hash>> hashes = {
+      {HASH_TYPE_MD5, std::make_shared<Hash>(HASH_TYPE_MD5)},
+      {HASH_TYPE_SHA1, std::make_shared<Hash>(HASH_TYPE_SHA1)},
+      {HASH_TYPE_SHA256, std::make_shared<Hash>(HASH_TYPE_SHA256)},
+  };
+
+  auto s = readFile(path,
+                    0,
+                    kHashChunkSize,
+                    false,
+                    true,
+                    ([&hashes, &mask](std::string& buffer, size_t size) {
+                      for (auto& hash : hashes) {
+                        if (mask & hash.first) {
+                          hash.second->update(&buffer[0], size);
+                        }
+                      }
+                    }),
+                    true);
```

Comment:
I think we talked about this in the past, and this is technically an issue on posix, as this flag is `blocking`. While you're here, would you mind updating this? I think we could drop something lile:
```
  auto blocking = isPlatform(PlatformType::TYPE_WINDOWS) ? true : false;
  auto s = readFile(path,
                    0,
                    kHashChunkSize,
                    false,
                    true,
                    ([&hashes, &mask](std::string& buffer, size_t size) {
                      for (auto& hash : hashes) {
                        if (mask & hash.first) {
                          hash.second->update(&buffer[0], size);
                        }
                      }
                    }),
                    blocking);

```

Reply:
So you want this to block for Windows only?

Reply:
I think we should isolate this change into another commit. So it's easily referenced later on.